src/utils/jsx-entity.js

```diff
@@ -0,0 +1,194 @@
+import { addComponent, addEntity, hasComponent } from "bitecs";
+import { preloadFont } from "troika-three-text";
+import {
+  CameraTool,
+  CursorRaycastable,
+  DestroyAtExtremeDistance,
+  FloatyObject,
+  HandCollisionTarget,
+  Holdable,
+  HoldableButton,
+  HoverButton,
+  MakeKinematicOnRelease,
+  Networked,
+  NetworkedTransform,
+  Object3DTag,
+  OffersHandConstraint,
+  OffersRemoteConstraint,
+  PhysicsShape,
+  RemoteHoverTarget,
+  Rigidbody,
+  SingleActionButton,
+  TextButton
+} from "../bit-components";
+import { inflateMediaFrame } from "../inflators/media-frame";
+import { inflateModel } from "../inflators/model";
+import { inflateSlice9 } from "../inflators/slice9";
+import { inflateText } from "../inflators/text";
+
+// TODO we should do this in a more explicit spot for "preloading" during the loading screen
+preloadFont({ characters: "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-_<()>[]|0123456789" }, function() {});
+
+function isValidChild(child) {
+  if (child === undefined) {
+    console.warn("found undefined node");
+    return false;
+  } else if (typeof child === "string") {
+    console.warn("found text node", child);
+    return false;
+  }
+
+  return true;
+}
+
+const reservedAttrs = ["position", "rotation", "scale", "visible", "name", "layers"];
+
+class Ref {
+  constructor() {
+    this.current = null;
+  }
+}
+export function createRef() {
+  return new Ref();
+}
+function resolveRef(world, ref) {
+  if (ref.current === null) {
+    ref.current = addEntity(world);
+  }
+  return ref.current;
+}
+
+export function createElementEntity(tag, attrs, ...children) {
```

Comment:
I was going for something like `React.createElement` here but I think this name is confusing. Open to suggestions. (This is the function JSX transforms to:
```js
<entity foo={{bar: 1}}>
  <entity biz />
</entiti>
```
translates to:
```js
createElementEntity("entity", {foo: {bar: 1}}, [
  createElementEntity("entity", {biz: true}}, null)
])
```